src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Summarize.cs

```diff
@@ -0,0 +1,245 @@
+                isValid = false;
+                errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrSummarizeDataSourceContainsThisGroupColumn);
+            }
+
+            var atLeastOneGroupByColumn = false;
+
+            for (int i = 1; i < args.Length; i++)
+            {
+                var argType = argTypes[i];
+                var arg = args[i];
+
+                DName columnName;
+                DType existingType;
+
+                // All args starting at index 1 need to be identifiers or aggregate functions (wrapped in AsNode node).
+                switch (arg)
+                {
+                    case AsNode nameNode:
+                        existingType = argType;
+                        columnName = nameNode.Right.Name;
+                        break;
+
+                    case FirstNameNode:
+                        if (!TryGetColumnLogicalName(sourceType, true, arg, errors, out columnName, out existingType))
+                        {
+                            isValid = false;
+                            continue;
+                        }
+
+                        if (DType.TryGetDisplayNameForColumn(sourceType, columnName, out var displayName))
+                        {
+                            if (displayName == FunctionThisGroupScopeInfo.ThisGroup)
+                            {
+                                isValid = false;
+                                errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrSummarizeThisGroupColumnName);
+                                continue;
+                            }
+                        }
+
+                        atLeastOneGroupByColumn = true;
+                        break;
+
+                    default:
+                        isValid = false;
+                        errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrSummarizeInvalidArg);
+                        continue;
+                }
+
+                if (columnName.Equals(FunctionThisGroupScopeInfo.ThisGroup))
+                {
+                    isValid = false;
+                    errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrSummarizeThisGroupColumnName);
+                    continue;
+                }
+                
+                // Restricted supported types
+                if (!(existingType == DType.String || 
+                      existingType == DType.Number || 
+                      existingType == DType.Decimal || 
+                      existingType == DType.Boolean || 
+                      existingType == DType.Time || 
+                      existingType == DType.OptionSet || 
+                      existingType == DType.OptionSetValue || 
+                      existingType == DType.DateTime ||
+                      existingType == DType.Date))
+                {
+                    errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrNeedPrimitive);
```

Comment:
>EnsureError [](http://example.com/codeflow?start=27&length=11)

Here we are enforcing that the grouping columns are primitive; do we also need to enforce that the grouping expressions return primitive values? For example, I don't think this should work: `Summarize(MyTable, Name, Filter(ThisGroup, Age >= 18))`

Reply:
Nevermind, found a test for that